STM32 Add PM Suspend to Ram support

gautierg-st opened the pull request:
This PR adds generic support for Suspend to RAM PM mode.

To that end, we use STM32 Standby mode with RAM retention (for SoC that provide it).
In low power applications, the Systick is typically given by a LPTIM. Since LPTIM is not clocked in Standby mode, before switching to this mode, we activate another clock (typically RTC) that works in Standby to play the role of Systick.
This was heavily influenced by the work in #63187 (Many thanks!).

A few peripherals (GPIO, ADC, Serial and entropy) have also been adapted to be compatible with this mode.

This other PR #67536 demonstrates the functionality on the STM32WBA series.

gautierg-st committed:
drivers: timer: stm32: lptim: remove init to 0 of static

Remove initialization of static variable to 0 to prevent resetting the
value when reinitializing the driver after resume from standby.
This has no impact since static variables are initialized to 0 by default.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: timer: stm32 lptim: add support for backup standby timer

Add support for a backup standby timer in STM32 LPTIM driver for cases
when the LPTIM is not available (ie standby low power mode).
A counter (typically RTC) is used for such a case.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: gpio: stm32: do not resume device when flag is DISCONNECTED

Fix a bug where after a standby, it was impossible to reenable a GPIO
clock.

A counter is incremented each time pm_device_runtime_get is called, and
decremented each time pm_device_runtime_put is called. The
clock is only enabled if this counter equals 1.
When configuring a GPIO (as input or output), the timer is incremented, and
when disconnecting it, it is both incremented and decremented. Thus the
next time we try to configuring it, the clock is not enabled (since the
counter will now be equal to 2).

This causes a problem when using low power standby mode: after wakeup all
clocks are disabled and the GPIO clock can not be reenabled.

This commit fixes this bug by not incrementing the counter when disconnect
is asked (or in other words incrementing it only when configuring either
an input or an output).

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: serial: stm32: add reinit after standby

When resuming from low power mode, if UART is disabled, this means that
we come from a mode that reset the registers, so we redo a full init of
the driver.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: stm32: prevent suspend to ram while measure in progress

Prevents the system to go in Suspend to RAM low power mode while ADC
measurement is in progress.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: entropy: stm32: prevent suspend to ram when operation in progress

Prevent the system to enter Suspend to RAM state while RNG operation is
in progress.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: entropy: stm32: reinit rng when exiting suspend to ram

With PM, when resuming from low power mode, reenable the RNG register
clocks and check the health register. If it is not set to the desired
value, it means we exit Suspend to RAM mode, and that the RNG needs to be
reinitialized.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

ceolin commented on drivers/adc/adc_stm32.c:
```diff
@@ -764,6 +764,8 @@ static void dma_callback(const struct device *dev, void *user_data,
 			adc_context_on_sampling_done(&data->ctx, dev);
 			pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE,
 						 PM_ALL_SUBSTATES);
+			pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_RAM,
+						 PM_ALL_SUBSTATES);
```
Nope, it just excludes the given state / substate. If the next scheduled event is enough for a deeper state be used, this will happen.

gautierg-st committed:
drivers: serial: stm32: prevent suspend to ram when operation in progress

Prevent the system to enter Suspend to RAM state while UART operation is
in progress.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st commented on drivers/timer/Kconfig.stm32_lptim:
```diff
@@ -56,4 +58,21 @@ config STM32_LPTIM_TICK_FREQ_RATIO_OVERRIDE
+config STM32_LPTIM_STDBY_TIMER
+	bool "Use an additional timer while entering Standby mode"
+	default $(dt_chosen_enabled,$(DT_CHOSEN_STDBY_TIMER))
+	depends on COUNTER
```
Done

gautierg-st commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -464,8 +527,6 @@ static int sys_clock_driver_init(void)
 	stm32_lptim_wait_ready();
 	LL_LPTIM_ClearFlag_ARROK(LPTIM);
 
-	accumulated_lptim_cnt = 0;
```
Done

gautierg-st commented on drivers/serial/uart_stm32.c:
```diff
@@ -2075,17 +2075,22 @@ static int uart_stm32_pm_action(const struct device *dev,
+				return err;
+			}
+		} else {
+			uart_stm32_init(dev);
```
Comment added

erwango reviewed (changes requested):
Otherwise LGTM

gautierg-st commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -524,9 +524,11 @@ static int gpio_stm32_config(const struct device *dev,
-	err = pm_device_runtime_get(dev);
-	if (err < 0) {
-		return err;
+	if (((flags & GPIO_OUTPUT) != 0) || ((flags & GPIO_INPUT) != 0)) {
```
With the current code, I don't think any problem can arisen, unless we use a mix of pin_ctrl_apply and gpio_pin_configure for the same GPIO, which looks like bad practice to me.
When using only gpio_pin_configure, if there is no clock, it means the GPIO was never configured first, so no problem.

gautierg-st commented on drivers/adc/adc_stm32.c:
```diff
@@ -764,6 +764,10 @@ static void dma_callback(const struct device *dev, void *user_data,
 			adc_context_on_sampling_done(&data->ctx, dev);
 			pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE,
 						 PM_ALL_SUBSTATES);
+#ifdef CONFIG_PM_S2RAM
```
Done
